src/filesystem.cpp

```diff
@@ -856,4 +853,207 @@ bool FileInputStream::read(byte *buffer, unsigned size)
+            localpath.append(localseparator);
+        #endif
+    }
+
+    localpath.append(additionalPath.localpath);
+}
+
+void LocalPath::separatorPrepend(const LocalPath& additionalPath, const string& localseparator)
+{
+    // no additional separator if there is already one after
+    if (localpath.size() >= localseparator.size() && memcmp(localpath.data(), localseparator.data(), localseparator.size()))
+    {
+        // no additional separator if there is already one before
+        if (additionalPath.editStringDirect()->size() < localseparator.size() || 
+            memcmp(additionalPath.editStringDirect()->data() + additionalPath.editStringDirect()->size() - localseparator.size(), localseparator.data(), localseparator.size()))
+        {
+            localpath.insert(0, localseparator);
+        }
+    }
+    localpath.insert(0, additionalPath.localpath);
+}
+
+void LocalPath::trimNonDriveTrailingSeparator(const FileSystemAccess& fsaccess)
+{
+    if (localpath.size() >= fsaccess.localseparator.size()
+        && !memcmp(localpath.data() + (int(localpath.size()) & -int(fsaccess.localseparator.size())) - fsaccess.localseparator.size(),
+            fsaccess.localseparator.data(),
+            fsaccess.localseparator.size()))
+    {
+        // ok so the last character is a directory separator.  But don't remove it for eg. F:\ on windows
+        #ifdef WIN32
+        if (localpath.size() > 2 * fsaccess.localseparator.size() && !memcmp(localpath.data() + localpath.size() - 2 * fsaccess.localseparator.size(), L":", fsaccess.localseparator.size()))
+        {
+            return;
+        }
+        #endif
+
+        localpath.resize((int(localpath.size()) & -int(fsaccess.localseparator.size())) - fsaccess.localseparator.size());
+    }
+}
+
+bool LocalPath::findNextSeparator(size_t& separatorBytePos, const FileSystemAccess& fsaccess) const
+{
+    for (;;)
+    {
+        separatorBytePos = localpath.find(fsaccess.localseparator, separatorBytePos);
+        if (separatorBytePos == string::npos) return false;
+        if (separatorBytePos % fsaccess.localseparator.size() == 0) return true;
+        separatorBytePos++;
+    }
+}
+
+bool LocalPath::findPrevSeparator(size_t& separatorBytePos, const FileSystemAccess& fsaccess) const
+{
+    for (;;)
+    {
+        separatorBytePos = localpath.rfind(fsaccess.localseparator, separatorBytePos);
+        if (separatorBytePos == string::npos) return false;
+        if (separatorBytePos % fsaccess.localseparator.size() == 0) return true;
+        separatorBytePos--;
+    } 
+}
+
+size_t LocalPath::getLeafnameByteIndex(const FileSystemAccess& fsaccess) const
+{
+    // todo: take utf8 two byte characters into account
+    size_t p = localpath.size();
+    p -= fsaccess.localseparator.size() % 2; // just in case on windows
+    while (p && (p -= fsaccess.localseparator.size()))
+    {
+        if (!memcmp(localpath.data() + p, fsaccess.localseparator.data(), fsaccess.localseparator.size()))
```

Comment:
yes, good spotting.  This is code that was moved from LocalNode::setnameparent , and I've changed it as little as possible otherwise.   Luckily there are only two cases - localseparator is 2 bytes on windows, and 1 bytes on linux.  I think it's ok to leave it as is on that basis, and because less change is less risky, what do you think?   thanks

Reply:
Agree. Let's keep it as it is. Thanks for double-checking :)